jtoh.go

```diff
@@ -49,40 +50,82 @@ func New(s string) (J, error) {
 }
 
 // Do receives a json stream as input and transforms it
-// in simple lines of text (newline-delimited) which is
+// in lines of text (newline-delimited) which is
 // then written in the provided writer.
 //
 // This function will block until all data is read from the input
 // and written on the output.
-func (j J) Do(jsonInput io.Reader, textOutput io.Writer) {
+func (j J) Do(jsonInput io.Reader, linesOutput io.Writer) {
 	jsonInput, ok := isList(jsonInput)
-	dec := json.NewDecoder(jsonInput)
+	// Why not bufio ? what we need here is kinda like
+	// buffered io, but not exactly the same (was not able to
+	// come up with a better name to it).
+	bufinput := bufferedReader{r: jsonInput}
+	dec := json.NewDecoder(&bufinput)
 
 	if ok {
 		// WHY: To handle properly gigantic lists of JSON objs
 		// Really don't need the return value, but linters can be annoying =P
 		_, _ = dec.Token()
 	}
 
+	var errBuffer []byte
+
 	for dec.More() {
 		m := map[string]interface{}{}
 		err := dec.Decode(&m)
+		dataUsedOnDecode := bufinput.readBuffer()
+		bufinput.reset()
+
 		if err != nil {
-			// TODO: handle non disruptive parse errors
-			// Ideally we want the original non-JSON data
-			// Will need some form of extended reader that remembers
-			// part of the read data (not all, don't want O(N) spatial
-			// complexity).
-			fmt.Fprintf(textOutput, "TODO:HANDLERR:%v\n", err)
-			return
+			errBuffer = append(errBuffer, dataUsedOnDecode...)
+			dec = json.NewDecoder(&bufinput)
+			continue
 		}
 
+		writeErrs(linesOutput, errBuffer)
+		errBuffer = nil
+
 		fieldValues := make([]string, len(j.fieldSelectors))
 		for i, fieldSelector := range j.fieldSelectors {
 			fieldValues[i] = selectField(fieldSelector, m)
 		}
-		fmt.Fprint(textOutput, strings.Join(fieldValues, j.separator)+"\n")
+		fmt.Fprint(linesOutput, strings.Join(fieldValues, j.separator)+"\n")
+	}
+
+	writeErrs(linesOutput, errBuffer)
+}
+
+func writeErrs(w io.Writer, errBuffer []byte) {
+	if len(errBuffer) == 0 {
+		return
 	}
+
+	// TODO: only do this in streaming mode
+	if errBuffer[0] == '\n' {
+		// Remove just the first \n since it is added
+		// because the newline after a valid document.
+		// Further newlines added on the output will be echoed.
+		errBuffer = errBuffer[1:]
+	}
+
+	// TODO: only do this in list mode
```

Comment:
All this list mode code feels very wrong... I think I'm going to remove this and just leave the output a little uglier when handling errors inside a list.